### ISSUE TITLE
Broken imports after moving files

### ISSUE THREAD

**Issue by yishaihl** [flydelabs/flyde#112]
Hey! Great job on the extension. I created a simple code flow and imported it into my flow, and it worked great. However, when I moved the imported file to a different folder, the import broke. I think it would be great if the extension could handle this case. Thanks

**Comment by Brijeshthummar02** [comment]
@yishaihl can you provide more details and short video of it if possible, i want to know more about the issue and give it a try.

**Comment by justinleeyang** [comment]
I can take bounty #112. I will share a scoped implementation plan and first PR shortly.

**Comment by justinleeyang** [comment]
Quick follow-up on #112: I am still actively working on this. Please let me know if you want me to narrow scope to a smaller first PR.

**Comment by justinleeyang** [comment]
No maintainer response yet on #112, I will switch to another open bounty for now and can resume this issue anytime when feedback arrives.

**Comment by Mira-Mjodheim** [comment]
/attempt #112

**Plan:**
- Perform deep trace of the reported issue in the local environment.
- Identify the root cause in the core logic.
- Implement a robust fix with regression tests.

**Comment by nexiumbiz-debug** [comment]
/attempt #112

Plan:
- Reproduce the broken-import case by creating/importing a flow, then moving the imported file to another folder.
- Trace how Flyde stores import references and how the extension resolves them after a file move.
- Patch import resolution/update logic so moved files keep working, with a focused regression test around relative paths.

Transparency note: this will be implemented with AI-assisted coding and verified before any PR. If the existing WIP already covers the same root cause, I will pivot rather than duplicate it.

**Comment by prashanthnimmagadda** [comment]
Submitted PR: https://github.com/flydelabs/flyde/pull/261

**Comment by b0r1v0j3** [comment]
/attempt #112

I have a focused loader fix with a regression case for relative imports from a moved visual node file.

**Comment by dipespandey** [comment]
/attempt #112

I am going to reproduce the moved-import case in the loader, identify where relative references are resolved from the wrong base path after a visual node is moved, then submit a focused patch with a regression fixture/test. I will keep the change scoped to import resolution and include local verification plus a short demo in the PR.

AI-assisted with Codex; I will review and verify the diff before submitting.

**Comment by ldthang0612** [comment]
/attempt #112

I have a focused VS Code extension fix ready that updates local .flyde import paths after file rename/move events. It covers both moving an imported flow and moving the importing flow, with targeted unit coverage and local compile/lint verification.

AI-assisted with Codex; I reviewed and verified the diff before submitting.

**Comment by dukunline-cyber** [comment]
/claim #112

Submitted PR: https://github.com/flydelabs/flyde/pull/266

Handles BOTH directions of the bug: a moved .flyde flow gets its own imports rewritten relative to the new location, AND moved import targets get all referencing flows updated workspace-wide. Standalone module + 4-line wire-up in extension.ts.

**Comment by Mira-Mjodheim** [comment]
Bonjour, je vais examiner l'issue. Pouvez-vous me donner plus de détails sur les fichiers déplacés et les erreurs d'import rencontrées ?

**Comment by Mira-Mjodheim** [comment]
Have you tried cleaning and rebuilding the project? That usually solves import issues caused by file relocations.

**Comment by nam157** [comment]
/attempt #112

I am adding a focused VS Code extension fix that rewrites relative file imports after .flyde file rename/move events. The patch covers both directions: updating imports inside a moved flow and updating other flows that reference a moved import target, with regression coverage for the path-rewrite helper.

**Comment by Mira-Mjodheim** [comment]
Could you please provide more details about the issues you're experiencing with broken imports after moving files? This will help me better understand and assist you in resolving the problem.

**Comment by nam157** [comment]
Thanks for checking in. The issue I reproduced is the one described in the original report: Flyde stores local visual-node imports as relative `source: file` paths, so those paths can become stale after files are moved in the VS Code workspace.

There are two concrete cases covered by PR #267:

1. Moving the importing `.flyde` file
   - Example before move:
     - `flows/main.flyde` imports `../nodes/child.flyde`
   - Move `flows/main.flyde` to `flows/nested/main.flyde`
   - Without rewriting, the import still points to `../nodes/child.flyde`, which now resolves from the new folder and breaks.
   - The PR rewrites that import so it still points to the same original target from the moved file's new location.

2. Moving the imported target `.flyde` file
   - Example before move:
     - `flows/main.flyde` imports `../nodes/child.flyde`
   - Move `nodes/child.flyde` to `shared/child.flyde`
   - Without rewriting, existing workspace flows still point at the old target location.
   - The PR scans workspace `.flyde` files and rewrites matching local file imports to the moved target's new relative path.

The implementation is scoped to the VS Code extension rename/move flow via `onDidRenameFiles`, with the path rewrite logic isolated in `vscode/src/flydeImportPathRewriter.ts` and regression coverage in `vscode/src/test/flydeImportPathRewriter.test.ts` for both directions above.

PR: https://github.com/flydelabs/flyde/pull/267
@Mira-Mjodheim

**Comment by ajjucoder** [comment]
/attempt #112

Plan: add VS Code rename handling for moved `.flyde` files. The patch rewrites relative file imports inside a moved flow and updates other workspace flows that reference a moved `.flyde` target, with focused regression coverage for both path-rewrite cases.

**Comment by Mira-Mjodheim** [comment]
I'd be happy to help with the broken imports issue. Can you please provide more details about the error messages you're seeing and the changes you made to the file structure?

**Comment by ajjucoder** [comment]
For context, the broken case is relative `.flyde` file imports after VS Code file moves.

PR #268 covers both file-structure changes:

1. Moving the importing flow
- Before: `flows/main.flyde` imports `../nodes/child.flyde`
- Move: `flows/main.flyde` to `flows/nested/main.flyde`
- Fix: rewrite the import so it still points to the same original target from the new folder.

2. Moving the imported target
- Before: `flows/main.flyde` imports `../nodes/child.flyde`
- Move: `nodes/child.flyde` to `shared/child.flyde`
- Fix: scan workspace `.flyde` files and rewrite references to the target moved path.

The change is wired through `vscode.workspace.onDidRenameFiles`, with regression coverage in `vscode/src/test/flydeImportPathRewriter.test.ts` for both directions.

PR: https://github.com/flydelabs/flyde/pull/268

**Comment by Mira-Mjodheim** [comment]
Hi, I'd be happy to help with the broken imports. Can you please provide more details about the files that were moved and the errors you're seeing? This will help me better understand the issue and provide a solution.

**Comment by nam157** [comment]
@Mira-Mjodheim For PR #267 specifically, there is not a separate build-time stack trace I am trying to address. The failure mode is that the `.flyde` file keeps a stale relative `source: file` import after a VS Code rename/move, so the editor/loader resolves the import from the wrong location and the referenced visual node can no longer be found.

A concrete file-structure example:

```text
Before:
workspace/
  flows/main.flyde          # imports ../nodes/child.flyde
  nodes/child.flyde
```

Case 1: move the importing flow:

```text
After:
workspace/
  flows/nested/main.flyde   # old import ../nodes/child.flyde is now stale
  nodes/child.flyde
```

From `flows/nested/main.flyde`, the import must be rewritten to `../../nodes/child.flyde` so it still points to the same target.

Case 2: move the imported target:

```text
After:
workspace/
  flows/main.flyde          # old import ../nodes/child.flyde is now stale
  shared/child.flyde
```

In that case, workspace flows that referenced the moved target need their local file imports rewritten to the target's new relative path.

The changes in #267 are limited to handling VS Code `onDidRenameFiles` events for `.flyde` files:

- rewrite imports inside the moved `.flyde` file relative to its new folder;
- scan other workspace `.flyde` files and rewrite references when the imported target was moved;
- keep the path rewrite logic isolated in `vscode/src/flydeImportPathRewriter.ts` with regression tests in `vscode/src/test/flydeImportPathRewriter.test.ts`.

So the "error" is the broken/stale local import path after file relocation rather than a new exception message introduced by my changes.

**Comment by Mira-Mjodheim** [comment]
Can you please provide more details about the issue you're experiencing with broken imports after moving files? What specific files were moved and what error messages are you seeing?

**Comment by nam157** [comment]
@Mira-Mjodheim The concrete files in PR #267 are local `.flyde` files using a relative `source: file` import. The regression tests model them as:

```text
workspace/flows/ParentFlow.flyde   # imports ./ChildFlow.flyde
workspace/flows/ChildFlow.flyde
```

Two moves are covered:

1. Move the importing file:

```text
workspace/flows/ParentFlow.flyde
-> workspace/flows/nested/ParentFlow.flyde
```

Before the fix, its stored import remains `./ChildFlow.flyde`, but from `workspace/flows/nested/ParentFlow.flyde` that points to `workspace/flows/nested/ChildFlow.flyde`, not the original `workspace/flows/ChildFlow.flyde`. PR #267 rewrites it to `../ChildFlow.flyde`.

2. Move the imported target:

```text
workspace/flows/ChildFlow.flyde
-> workspace/shared/ChildFlow.flyde
```

Before the fix, existing flows still point to `./ChildFlow.flyde`. PR #267 rewrites the reference from `ParentFlow.flyde` to `../shared/ChildFlow.flyde`.

There is no new custom error string from my change. The user-visible failure is the existing broken import behavior after relocation: Flyde resolves the stale relative path from the file's current folder, so the referenced visual node cannot be found/loaded. The fix is intentionally scoped to VS Code `onDidRenameFiles` handling plus the pure path rewriter in `vscode/src/flydeImportPathRewriter.ts`, with coverage in `vscode/src/test/flydeImportPathRewriter.test.ts`.

**Comment by Catnap7** [comment]
/attempt #112

Submitted PR for this bounty: https://github.com/flydelabs/flyde/pull/271

This focuses on keeping local `.flyde` imports valid after imported flow files/folders are moved in VS Code, with a small pure helper plus unit coverage.

**Comment by Mira-Mjodheim** [comment]
Merci pour cette notification ! Pouvez-vous fournir plus de détails sur les imports qui sont cassés et les fichiers qui ont été déplacés ? Cela m'aidera à mieux comprendre le problème et à trouver une solution.

**Comment by Catnap7** [comment]
@Mira-Mjodheim The concrete broken case covered by PR #271 is a local `.flyde` import stored as a relative `source: file` path after a VS Code file/folder move.

Example covered by the tests:

```text
workspace/flows/Parent.flyde    # imports ./Child.flyde
workspace/flows/Child.flyde
```

If `Child.flyde` is moved to `workspace/shared/Child.flyde`, the existing import in `Parent.flyde` still points to `./Child.flyde`, so it resolves to the old location and the imported visual node can no longer be found. PR #271 rewrites references to the moved file so `Parent.flyde` points to `../shared/Child.flyde` instead.

The implementation is intentionally scoped to VS Code rename handling for imported `.flyde` files/folders, with path rewrite logic in `vscode/src/flydeImportPathRewriter.ts` and unit coverage in `vscode/src/test/flydeImportPathRewriter.test.ts`.

**Comment by Mira-Mjodheim** [comment]
Thanks for the ping! I'll take a look at the broken imports and investigate the issue.

**Comment by Catnap7** [comment]
Submitted PR #271 for this bounty: https://github.com/flydelabs/flyde/pull/271

The PR updates local .flyde imports when VS Code reports file/folder moves, including moved imported flow files, moved containing folders, and moved parent flows while leaving package/custom sources untouched.

Verification on the PR: pnpm compile-tests and focused mocha coverage for the import path rewrite behavior.

**Comment by arancione3000** [comment]
/attempt #112

I’m taking this issue and will trace the import-path handling, add a focused regression test for moving an imported flow to another folder, and submit a minimal fix if the behavior is reproducible.

**Comment by andrewscar34-ux** [comment]
/attempt #112

I have prepared a fix using VS Code’s onWillRenameFiles event to update file references through a WorkspaceEdit. It handles dependency moves, importing-flow moves, folder renames, and nested inline flows. Nine focused regression tests cover path handling, including simultaneous moves.

The implementation, tests, and build are ready; I will submit the PR with a short presentation of the actual before/after extension-host test results and the reproducible verification harness. Prepared and tested with OpenAI Codex assistance.

**Comment by saphid** [comment]
AI-assisted scope audit before attempting #112: I inspected the issue, contribution rules, and the actual diffs/discussion for #256, #259-#271, and #273-#276. I also fetched and tested the strongest current candidates. PR #275 already appears to cover producer, consumer, file/folder, batch, inline-node, dirty-buffer, and package/custom cases, with claimed VS Code host/demo evidence, so I do not want to add another duplicate `/attempt`.

One reproducibility note: on macOS, #276 compiles and its 11 focused tests pass, but its submitted real extension-host suite finishes with 24 passing / 1 failing. The rename integration test registers a handler while the development extension has already registered the same handler, producing overlapping WorkspaceEdit ranges; this is a test/validation issue, not evidence of a separate product requirement.

Could a maintainer confirm whether #112 requires any behavior not already represented by #275/#276 (for example, recovery after moves made outside VS Code rather than VS Code-originated rename events), and whether the advertised Algora bounty is still funded and claimable? I am not requesting assignment, claiming the bounty, or asserting payout eligibility. If there is no additional intended scope, I will leave this to the existing contributors rather than submit another implementation.

**Comment by BreakYukino** [comment]
/attempt #112

I am working on a focused VS Code rename handler that updates local file references when a referenced node or its parent folder moves. It will also rebase relative references when the importing flow moves. I plan to use a WorkspaceEdit from the rename event so unsaved editor buffers and undo stay under VS Code's control, preserve YAML formatting, and cover file/folder moves plus nested inline nodes with regression tests.

This is an AI-assisted contribution (Codex) on behalf of the account owner. I will include verification evidence and the requested demonstration with the submission.

**Comment by bobbyhuang-dev** [comment]
Hi — is the advertised $15 bounty for imports breaking after a file move still available to a new contributor? I see existing attempts and want to avoid duplicating them. A submission would use AI-assisted development and include a regression test and demo video. Could you confirm current eligibility, reward availability and whether a completed fix can be reviewed within 48 hours? If this issue is already covered, no action needed.

**Comment by saurav-codes** [comment]
/attempt #112

Plan: add a rename/move watcher in the VS Code extension that rewrites relative file import paths in .flyde flows (both the moved import target and a moved flow's own imports), using a YAML document transform plus regression tests.

**Comment by wangfumin1** [comment]
/attempt #112

Plan:
- Reproduce the broken import after moving an imported flow/file between folders.
- Trace how Flyde resolves and persists import paths, then update path resolution so moved files keep valid references.
- Add regression coverage for the move scenario and verify existing import behavior remains intact.
- Submit the required demo evidence with the PR.

**Comment by kestrel-devagent** [comment]
/attempt #112

Plan: reproduce broken relative imports after moving a `.flyde` flow between folders in the VS Code extension; implement/adjust rename path rewriting for imported flows and dependent imports; add a focused regression test; open a minimal PR.

**Comment by kestrel-devagent** [comment]
/claim #112

Submitted PR: https://github.com/flydelabs/flyde/pull/282

VS Code `onWillRenameFiles` handler rewrites local `source: file` imports (and deprecated top-level import keys) when `.flyde` files/folders move, including when the importing flow itself moves. Unit coverage for both directions, folder moves, inline nodes, and batch renames.

**Comment by adnandispatch9-jpg** [comment]
Hi maintainers — I reviewed issue #112 and the existing file-rename implementation in PR #282. Before starting overlapping work, could you confirm whether the advertised $15 bounty is still available and identify any specific acceptance criterion not covered by the existing PRs? I would only proceed with a distinct, agreed scope and include regression tests plus the required demo. Any contribution from this account would be AI-assisted. This is a scope/funding clarification, not an attempt or bounty claim.

**Comment by huipan18** [comment]
Hi! I've successfully reproduced this issue locally on Windows and reviewed PR #275.

I noticed that a fix has already been submitted. Is there any remaining work, additional testing, or an alternative approach that would be helpful?

I'd be happy to contribute. Thanks!

**Comment by themaksat** [comment]
I have submitted a solution for this issue in PR https://github.com/flydelabs/flyde/pull/284. Full implementation and verification included. Payout wallet (Solana): `2BTCUUTviLNDtZDRBhPBTwSXZMrTttPJNbGevMXBBMKK`

**Comment by projectcarbonfiber** [comment]
/attempt #112

Working on a dual-path fix for broken file imports after moves:
1. VS Code `onWillRenameFiles` + scalar-only YAML rewrite in `@flyde/loader` (Explorer renames)
2. Loader basename-search recovery when the stored path is already stale (external/fs moves)

Focused hermetic tests included. PR incoming.
